### DISCUSSION TITLE
how to improve perf with multiple topics per socket

### DISCUSSION THREAD

**Discussion by rpgehlot** [uNetworking/uWebSockets.js#553] [Q&A] [answered]
Hello, I have been using uwebsocket in production environment under two different use cases.

**Use case:1** Number of clients are larger but topics are much less, so using pub-sub works well here. So this use case is successful with uws since the underlying idea of maintaining topictree in uws and batching updates together for a event loop iteration yields good results.

**Use case:2** Number of clients are smaller( ~20 ) but they have little high number of unique topics per socket they subscribe to which can be around ~10 as a rough number. But here i see huge lag in updates like minutes which basically seems to be happening becuase of larger number of overall topics.

the data input is same for both use cases. Cpu uasge is constantly 100% for second use case
topics are something like protocol/machine/team/account/positions
Is there any workaround through which we can improve perf in second use case ? would iterating over sockets for a topic and doing .send() call a better option here ?
Thanks.

**Comment by rpgehlot** [comment]
can anyone give any hints/ideas regarding this ?

**Reply by rpgehlot** [reply]
yes i read them, from what i understand it makes sense to use a for loop over sockets subscribed to a topic and call `.send(buf, true)` on each one of them, that would be faster in case if you have multiple topics per socket  since using pub-sub has a higher cost associated with it for such a case. but i guess that will have a upper bound since it makes it quadratic in time complexity, maybe we can just create a cluster of uws server to scale it up thereafter.
pls correct me if am wrong.

**Reply by rpgehlot** [reply]
If i somehow reduce topics to **one per socket**, that should improve perf right ? as said on discussion #558

**Reply by rpgehlot** [reply]
i think this flag `nonStrict` is related to performance of subscription and unsubscription from topics. how would it affect the performance of publish its unclear to me.

**Reply by rpgehlot** [reply]
ok, let me rephrase it again, when there are 9-10 subscribed topics per socket and socket count is pretty small, i see significant delay in a packet reaching from uws server to client browser which is handled using `.publish(topic, buf, true)` 
while the same delay is not there in other use case when many sockets just subscribes to 1 single topic, so there are actually only a handful of topics but a large number of sockets here which is also handled using publish.
So my question is how would `nonStrict` flag help in bringing down the latency in the former case since it is related to subscribe/unsubscribe ?

**Reply by rpgehlot** [reply]
also topic is of a form `a/b/c`, can that cause perf problem ? or is the problem just related to number of topics subscribed to a socket ?

**Reply by hst-m** [reply]
nonStrict would only affect subscribe and unsubscribe performance so is not relevant. So you are publishing to 200 topics for 20 clients 10 topics each or is that 10 topics total

**Reply by rpgehlot** [reply]
200 for 20 clients 10 topics each

**Reply by rpgehlot** [reply]
So, I used some technique through which i reduced number of overall different topics equal to number of websockets client in worst case, 
for e.g. a `client A `was subscribed to set of topics `{ t1, t2, t3 }`, so i mapped this group to a single `virtual topic tA` and pushed any data for `{ t1, t2, t3 }` into tA. So essentially i was trying to keep number of triggered topics on `uws level` equal to number of clients in worst case and since i had only 20 clients, it resulted in 20 different overall triggered topics in worst case in a single event loop cycle. But still perf was suffering, maybe 20 overall topics is also too high. what is the ideal number of overall triggered topics here ?

**Reply by rpgehlot** [reply]
is there some option to increase this upper limit of 64 topics here like is it configurable ?

**Reply by ghost** [reply]
There is no upper limit of 64 topics, you can have millions of topics if you want to but the gain for using pub/sub over regular send is minimal if not negative in those cases. What the 64 limit is, is more an ideal number of simultaneously active topics. So ideal cases would have a handful of topics being shared among many sockets, rather than a buttload of topics.

**Reply by ghost** [reply]
What platform are you on?

**Reply by rpgehlot** [reply]
I am using this on centos 7, as for ideal case, it is working well for me as well.

**Reply by ghost** [reply]
Minutes of lag is a hint that something is seriously wrong in your application - latency should be micro- or milliseconds. Given that you have asked regarding 11 MB of backpressure before, I'm assuming you're still doing something inherently wrong here. The fact that the problem manifests using the pub/sub does not necessarily mean there is something wrong with pub/sub, it could very well just mean it happens to show the problem sooner.

**Reply by rpgehlot** [reply]
I am not saying there is something wrong with pub sub, it definitely has much lower latency in comparison to other libraries that I have used and i roughly understand how it works, which is evident to me in the ideal case. In the other case where each user has a group of different topics, this problem is showing up which is what i wanted to check if it can be improved. 11MB was reduced to 5MB back pressure eventually to send a bigger data packet at start of connection using pub sub, which shouldn't be the issue as it is working fine in other case as well. In the app I am not doing anything extraordinary, just that instead of publishing data in string format I am encoding it to a smaller size packet and calling .publish per sec to send all accumulated data at once.

**Reply by e3dio** [reply]
> calling .publish per sec to send all accumulated data at once.

Have you tried sending data as it comes in instead of large batch every second

**Reply by ghost** [reply]
It's definitely optimized for only a few topics (50-ish or less) and a lot of connections (very high ratio of shared topics). That's where it excels. For everything else where you have a lot less topic sharing, just use ws.send.

At least that is how it is right now, I might add other implementations with different strategies later on so that users can compile or swap strategy as needed.

So yeah, definitely best for limited number of topics and tons of users.

**Reply by ghost** [reply]
It only prepares the combinations in use, so worst case is one combination per socket. But if the developer keeps to the design, using topics that are shared, the probability of sharing combinations increase with increased number of sockets.

**Reply by ghost** [reply]
It affects more than this, so a complete copy + paste is probably needed

**Reply by rpgehlot** [reply]
+1 for supporting different publish strategies, based on use cases. User can specify if they want to use standard pub-sub or pub-sub strategy optimised for handful topics , but larger number of sockets while initialising the uws instance. I am not sure how difficult it would be, but can definitely useful for certain kind of cases.

**Reply by uasan** [reply]
In one project, there can be several strategies at once, and some topics can change their strategy in real time, in the morning one topic for a small number of users, in the evening this topic for a large number of users and vice versa.
Strategies should be built on statistics of subscribers in real time, the statistics are simple, what percentage or number absolute of users are subscribed to this topic.

**Reply by ghost** [reply]
You can have two levels - "isolation" and "topic" where isolation goes to a different tree. Then you can easily do things like multi tenant stuff and if topics are naturally isolated you just use a ton o isolations. Either way, this can be a good addition to the API

**Reply by ghost** [reply]
But even in worst case, since you have 64 topics per batch means you can do at least 64 publishes per loop iteration, and if you publish to the same 64 topics you can publish how much you want. Still, having 64 publishes per event loop iteration is quite a lot and the pub/sub shouldn't completely halt for minutes - and again, you have reported having 11MB of backpressure so I would start looking at your usage rather than blaming the pub/sub. It can be improved, definitely, and it does perform less than great in certain cases but it shouldn't be a blocker for making a working app.

**Reply by ghost** [reply]
I'm working on a benchmark for worst and best cases of current algorithm and from what I can see, 100k subscribers in worst case (they all subscribe to their own topic AND they are senders) takes 900ms to process 100k publishes.

Best case, or what it is built for, when you have 100k subscribers to a maximum of 64 topics INCLUDING having senders takes 170ms to process 100k publishes.

So current algorithm is pretty okay even for worst possible case, sure, it's going to emit tons of individual packets to the networking layer but that's not worse than what other crappy pub/subs do anyways. So worst case isn't really that horrifying.

If you see minutes of 100% CPU-time you are most probably just shoving way too much data through the system, like your previous bug reports point to.

**Reply by ghost** [reply]
I mean it's kind of self explanatory; if you put 1 topic for every single socket, then do a loop over all topics publishing 1 message to all topics, 1 a time, then you're obviously going to get the worst possible utilization of pub/sub. That just mimics using a loop with send!

I don't really see the problem, this algorithm seems to perform even better than what I anticipated - there is only a 5x difference between best possible case and worst possible case.

**Reply by ghost** [reply]
@rpgehlot fix your app. You're not going to see any improvement by waiting for changes to uWS

**Reply by ghost** [reply]
Note that the worst possible case tested is having **100k topics and 100k subscribers** so clearly the algorithm has no problem dealing with tons of topics - it's just that the most optimal drainage of published messages happen when you publish to topics that have high ratio of sharing. But even if you would publish to topics that have only 1 socket, you can still get okay performance if you publish all messages that belong to this topic **before moving on to other topics**.

So as long as you don't have a usage that naturally maps to a for loop (for 10 messages; for every socket; publish a message), there should be no issues at all. And in that case, well in that case you basically have created the most idiotic app anyways.

**Reply by ghost** [reply]
The rule of thumb for publish is basically the same rule of thumb as for send; publish/send to topics/with sockets in ways to **maximize locality** - if you have 10 messages to send, send them to **one socket a time**, then move on to next. Same for publish; **publish to topics in order of locality / relationship** - use the sockets/topics that are the most related, in groups, don't just randomly publish.

**Reply by Shonke** [reply]
> There is no upper limit of 64 topics, you can have millions of topics if you want to but the gain for using pub/sub over regular send is minimal if not negative in those cases. What the 64 limit is, is more an ideal number of simultaneously active topics. So ideal cases would have a handful of topics being shared among many sockets, rather than a buttload of topics.

Sorry to reply to closed discussions.
In your opinion, I should use fewer topics and broadcast to more clients.

Now I want to use multiple processes. I want to send the same topic to the same process for processing. That's why I want it.
https://github.com/uNetworking/uWebSockets.js/issues/649

Do you have any suggestions about this?

**Reply by Shonke** [reply]
Or do I not need multiple processe to handle this? Will uWebSockets use all CPU cores?

**Reply by e3dio** [reply]
This discussion is outdated, v20 Publish is updated to do 65,536 unique messages per round instead of 64 unique topic combinations, topics are lightweight and using many is fine

**Reply by Shonke** [reply]
@e3dio Thank you for your answer.
Number of processes and CPU cores, is there any suggestion?

**Comment by ghost** [comment, marked as answer]
Repeating the conclusion:

The rule of thumb for publish is basically the same rule of thumb as for send; publish/send to topics/with sockets in ways to maximize locality - if you have 10 messages to send, send them to one socket a time, then move on to next. Same for publish; publish to topics in order of locality / relationship - use the sockets/topics that are the most related, in groups, don't just randomly publish.

**Reply by ghost** [reply]
LOL! My benchmark was run with ASAN and without optimization - with optimization the numbers are:

```
TestPerformance
Testing best case
41 ms
59 ms
68 ms
68 ms
67 ms
68 ms
69 ms
69 ms
69 ms
70 ms
1000000
Testing worst case
130 ms
134 ms
131 ms
128 ms
134 ms
134 ms
135 ms
131 ms
134 ms
133 ms
```

In other words; there is absolutely nothing wrong with the current algorithm; it beats my expectations by a lot.

**Reply by ghost** [reply]
The best addition would be to simply add isolations. Then you can do multi tenant stuff like said, where it is not really possible to organize publishes by locality, since they are isolated in nature. And you don't have to use it for only multi tenant but can even use it in many layers like tentant -> group -> topic by having the isolation be "tenant/group". That gives a lot of power to structure things really efficiently.

**Reply by e3dio** [reply]
The selected answer should be updated for v20 Publish, see https://github.com/uNetworking/uWebSockets.js/discussions/553#discussioncomment-2414724

**Reply by Shonke** [reply]
> The selected answer should be updated for v20 Publish

It'll be in the document, won't it?

**Comment by e3dio** [comment]
.

**Reply by ghost** [reply]
Alright I did like this:

100k subscribers. 17 topics. Every subscriber subscribes to all the bits set in an increasing 17-bit counter (that is, all subscribers have entirely unique set of topics, yet will subscribe to more than only 1 topic).

This is what I got, after publishing once to all 17 topics (10 times over):

```
TestPerformance
Testing worst possible case
intersectionCache size is: 99999
286 ms
intersectionCache size is: 99999
198 ms
intersectionCache size is: 99999
196 ms
intersectionCache size is: 99999
196 ms
intersectionCache size is: 99999
198 ms
intersectionCache size is: 99999
199 ms
intersectionCache size is: 99999
201 ms
intersectionCache size is: 99999
199 ms
intersectionCache size is: 99999
199 ms
intersectionCache size is: 99999
198 ms
```

You can see the size of intersectionCache is 100k, the worst possible level of sharing (nothing shared), yet it only took 200ms. This surely is the absolute worst possible case you can do. No matter how many combinations you can have, intersectionCache can only hold less than 100k, since it lazily fills with the combination in use.

**Reply by ghost** [reply]
A quick and simple optimization makes it like so:

```
Testing worst possible case
intersectionCache size is: 99999
230 ms
intersectionCache size is: 99999
149 ms
intersectionCache size is: 99999
148 ms
intersectionCache size is: 99999
150 ms
intersectionCache size is: 99999
148 ms
intersectionCache size is: 99999
149 ms
intersectionCache size is: 99999
150 ms
intersectionCache size is: 99999
150 ms
intersectionCache size is: 99999
149 ms
intersectionCache size is: 99999
149 ms
```

If you remove the quick-fix of holes (v19) it becomes:

```
Testing worst possible case
intersectionCache size is: 99999
194 ms
intersectionCache size is: 99999
126 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
128 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
127 ms
intersectionCache size is: 99999
127 ms
```

If you only have one "data track", or just optimize the append, it becomes:

```
Testing worst possible case
intersectionCache size is: 99999
125 ms
intersectionCache size is: 99999
95 ms
intersectionCache size is: 99999
101 ms
intersectionCache size is: 99999
97 ms
intersectionCache size is: 99999
96 ms
intersectionCache size is: 99999
99 ms
intersectionCache size is: 99999
102 ms
intersectionCache size is: 99999
96 ms
intersectionCache size is: 99999
96 ms
intersectionCache size is: 99999
97 ms
```

**Reply by ghost** [reply]
If you just emit the data without collecting it in intersectionCache (so having to use the cork buffer all the time), it gets like this:

```
TestPerformance
Testing worst possible case
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
47 ms
intersectionCache size is: 0
49 ms
intersectionCache size is: 0
47 ms
```

But now you have kind of optimized for worst case, and the best case is no longer nicely "cached".

**Reply by ghost** [reply]
I should make a best case benchmark where only 1 single topic is used for all sockets!

**Reply by ghost** [reply]
I mean it's really about where you do the collecting of messages into one; either you do it as part of intersectionCache, that is, you store the collected messages in a place where it can be accessed again without re-collecting the same messages (which obviously is best for shared topics), or you remove intersectionCache and pack, collect, the messages into one for every single socket (hoping for minimal sharing).

Either way you have to pack the different messages into one string, for every socket.

Having intersectionCache is quite fast (it is a red-black tree with O(log n) insert in worst case, but yeah it can hold a lot of duplicated data, however this is already something commonly done in other solutions:

A common solution to pub/sub is to have every socket hold its own user space buffer, where data is appended to (causing as much duplicated data to be held per socket as the intersectionCache). Then you finally copy that to kernel buffers.

So current cache is not really that much worse (in worst case) than what other pub/sub solutions are in average case. This while being much better in the cases where you do have shared topics (such as when all sockets are subscribed to 1 topic and you publish to that topic).

Overall, this implementation needs to be refactored, benchmarked and optimized and yes there can come more implementations but my benchmarking of it, simply shows that it is fairly okay already, even in worst case.

**Reply by ghost** [reply]
This code really needs to be rewritten, but I like the idea of it. 64 is not even something you need to limit it to - you can simply replace that for std::bitset<512> or whatever, and it works just fine. Same performance. So that raises the max topics triggered in one batch to 512.

**Reply by ghost** [reply]
It just happens to be ordered, there's still no guarantee (it can be changed later). The main benefit of sorting them like that is to remove duplicates (such as when subscribing to overlapping topics like wildcards)
